### ISSUE TITLE
MAC/Cycle mismatch

### ISSUE THREAD

**Issue by Sixaxis9** [MaximIntegratedAI/ai8x-synthesis#136]
Hi,
I have been using the chip MAX78000 for quite some time now, and I am really impressed by the product.

However, I cannot explain why I have an inference time of less than 1ms for a 5 million MACs network. I got the number of MACs from the file cnn.h created with the synthesis tool, and the 1ms has been measured with an oscilloscope on the LED2 (red led on the ev board), that should light up when the network is running.

The accelerator is running at 50MHz, and thus it takes 50.000 cycles to complete the inference, and the result is correct and passes the check. But that would mean that the accelerator would do 5e6/5e4 = 100 MAC/Cycle, and according to the datasheet "Nominal 1 output channel per clock", I think the maximum theoretical value of MAC/Cycle should be 64, equal to the number of processors.

I am trying to understand this behaviour, do you think this is normal? Am I right taking into account a max of 64MAC/Cycle? My take is that the accelerator is not really running my network...
[neural_net.zip](https://github.com/MaximIntegratedAI/ai8x-synthesis/files/6587606/neural_net.zip)

Thank you for your time!

**Comment by rotx-maxim** [comment]
*TLDR*: The hardware can write 32-bits of output per cycle (four channels), and therefore your maximum rate is off by a factor of 4.

A little more detail:

First of all, your time measurement is approximately correct. I recreated your network (using random data and random weights) and get a bit less than 1 ms just using the built-in timer.

Second, the hardware can write one 32-bit output word per cycle (or four 8-bit results).
Layers that don’t use all 64 processors obviously aren’t as efficient, that’s why it’s taking almost one millisecond (when in theory this should be doable in less time).

Here’s a breakdown of the approximate time for each layer. The time is measured using the ARM timer, so not perfect.

Layer	|  MACs	| approx time (µs)
---- | ------------ | ---------------------------
0	|    225,792	| 545	NOTE: least efficient due to single processor
1	| 4,713,824	| 358
2	|    585,952	|   76
3	|      46,656	|     7
4	|         6,912	|     7
5	|             512    |   <1
6	|              80    |   <1

For more exact measurements, you would use the EVkit and its power accumulator.

A passing known-answer test is usually a pretty good indication of whether the network is actually running. Unless the output is all zeros, so check the log file (log.txt) for all intermediate values to make sure.

**Comment by Sixaxis9** [comment]
Thanks a lot rotx-maxim! That clarifies a lot.
So it would be correct to expect a maximum throughput of 64 (processors) * 4 (8 bit data per processor) per clock cycle?
Following your time breakdown, I would expect 545us*50MHz = 27.2k cycles for the first layer, but if I consider a throughput of 4 bytes per processor, this would lead to a bit more than to 100k MACs, while the layer involves 225k MACs. Same for layer 2, 17.9k cycles, which considering 32 processors would lead to 17.9k*4*32 = 2.2M. I am still about a factor 2 off...
Am I missing something there?
Again, thanks a lot for clarifying this aspects. I really enjoy the product and I am looking forward to extract the maximum performance from it :)

**Comment by rotx-maxim** [comment]
The first layer is 1x30x30 in and 32x28x28 out with a non-padded conv2d and 3x3 kernel. The number of MACs should be kernel_size * input_channels * output_channels * output_dim = 3*3 * 1 * 32 * 28*28 = 225,792, so we have the correct number for this layer.

We need a minimum of 32 * 28*28 / 4 = 6,272 cycles to write the data. So this number is not the bounding factor here and I apologize for any confusion (edited first reply).

We don't have the latency calculations documented at this time.
For this simple example, it is roughly output_dim * (output_chan + 2) = 28*28 * (32 + 2) = 26,656 cycles or ~533 us. This is close to the number measured by the timer.

As you can see, this is a little less than 9 MACs per cycle which makes sense given the 3x3 kernel size and the fact we can consume a full kernel every cycle, but also have to fetch and write data which mostly happens in parallel.

We will add some clarification to the user guide.
